### PR TITLE
Improve spec compliance of discarding BCs

### PR DESCRIPTION
**Pull request by gterzian**
Came-up at https://github.com/servo/servo/pull/23637#issuecomment-528254988

---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/24143)

### CODE REVIEW

**Comment by gterzian** [comment]
@jdm @asajeffrey r?

The problem is still running tasks after either a frame has been removed from the tree, or an auxiliary closed, in both cases the BC is discarded "sync"(as per the spec), however tasks are only ignored later when the script-thread handles the "exit pipeline" message, which means that in the meantime the global can still be handling tasks. This became apparent in https://github.com/servo/servo/pull/23637#issuecomment-528254988

**Comment by gterzian** [comment]
For good measure I'm also preventing handling any user input via compositor events, since those result in firing events and potentially running JS code, if the BC is being discarded for a top-level, like an auxiliary(not an iframe). 

Reason for this is the [spec says](https://html.spec.whatwg.org/multipage/window-object.html#discard-a-document) to "Remove any tasks associated with document in any task source, without running those tasks.", and so running event handlers seems to go against that(I'm actually surprised those event handlers are not run by way of a task).  

I assume iframe do not handle compositor events, right?

**Comment by gterzian** [comment]
well, look at that...

```
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "worker-src 'self' directive should disallow cross origin static import.", 
    "test": "/workers/modules/dedicated-worker-import-csp.html", 
    "line": 284199, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "script-src 'self' directive should disallow cross origin static import.", 
    "test": "/workers/modules/dedicated-worker-import-csp.html", 
    "line": 284201, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "worker-src 'self' directive should override script-src * directive and disallow cross origin static import.", 
    "test": "/workers/modules/dedicated-worker-import-csp.html", 
    "line": 284204, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "script-src 'self' directive should disallow cross origin dynamic import.", 
    "test": "/workers/modules/dedicated-worker-import-csp.html", 
    "line": 284205, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "OK", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/workers/modules/dedicated-worker-import-csp.html", 
    "line": 284211, 
    "action": "test_result", 
    "expected": "CRASH"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "promise_test: Unhandled rejection with value: \"FAIL - first document not exposed. navigated_back is undefined\"", 
    "stack": null, 
    "subtest": "Test that iframe navigations are not observable by the parent, even after history navigations by the parent", 
    "test": "/resource-timing/nested-context-navigations-iframe.html", 
    "line": 303458, 
    "action": "test_result", 
    "expected": "TIMEOUT"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "promise_test: Unhandled rejection with value: \"FAIL - first document not exposed. navigated_back is true\"", 
    "stack": null, 
    "subtest": "Test that crossorigin iframe navigations are not observable by the parent, even after history navigations by the parent", 
    "test": "/resource-timing/nested-context-navigations-iframe.html", 
    "line": 303459, 
    "action": "test_result", 
    "expected": "NOTRUN"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "promise_test: Unhandled rejection with value: \"FAIL - initial document should be observable\"", 
    "stack": null, 
    "subtest": "Test that iframe navigations are not observable by the parent", 
    "test": "/resource-timing/nested-context-navigations-iframe.html", 
    "line": 303460, 
    "action": "test_result", 
    "expected": "NOTRUN"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "promise_test: Unhandled rejection with value: \"FAIL - initial document should be observable\"", 
    "stack": null, 
    "subtest": "Test that crossorigin iframe navigations are not observable by the parent", 
    "test": "/resource-timing/nested-context-navigations-iframe.html", 
    "line": 303461, 
    "action": "test_result", 
    "expected": "NOTRUN"
}
{
    "status": "TIMEOUT", 
    "group": "default", 
    "message": "Test timed out", 
    "stack": null, 
    "subtest": "Test that iframe refreshes are not observable by the parent", 
    "test": "/resource-timing/nested-context-navigations-iframe.html", 
    "line": 303462, 
    "action": "test_result", 
    "expected": "NOTRUN"
}
{
    "status": "TIMEOUT", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/resource-timing/nested-context-navigations-iframe.html", 
    "line": 303464, 
    "action": "test_result", 
    "expected": "CRASH"
}
```

And also 

```
{
    "status": "CRASH", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/html/browsers/the-window-object/close-method.window.html", 
    "line": 172617, 
    "action": "test_result", 
    "expected": "OK"
}
```

**Comment by gterzian** [comment]
```
{
    "status": "NOTRUN", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "Test that crossorigin iframe refreshes are not observable by the parent", 
    "test": "/resource-timing/nested-context-navigations-iframe.html", 
    "line": 303982, 
    "action": "test_result", 
    "expected": "TIMEOUT"
}
``` 

This one seems to intermittently alternate between TIMEOUT and NOTRUN...

**Comment by gterzian** [comment]
```
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "window.close() queues a task to discard, but window.closed knows immediately", 
    "test": "/html/browsers/the-window-object/close-method.window.html", 
    "line": 182482, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "NOTRUN", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "Test that crossorigin iframe refreshes are not observable by the parent", 
    "test": "/resource-timing/nested-context-navigations-iframe.html", 
    "line": 159318, 
    "action": "test_result", 
    "expected": "TIMEOUT"
}
```

**Comment by gterzian** [comment]
@asajeffrey r?

**Comment by gterzian** [comment]
I moved the `is_closing` flag from the window to the proxy, since it's an attribute of the BC. https://html.spec.whatwg.org/multipage/browsers.html#is-closing

**Comment by gterzian** [comment]
The final piece of the puzzle, I think, to be done in a follow-up, is to not close the pipeline of an iframe when it's still reachable from script and potentially re-inserted into the document. So we should discard the nested BC when the frame is removed, but not the pipeline. And then if the frame is added to the document again, we should create a new BC but re-use the same pipeline, or something similar. 

The trick could be keeping a `Weak` to the iframe, and only closing the pipeline if the frame has been GC'ed, since at that point we can be sure it can never be re-inserted into the doc.

File: components/script/dom/document.rs [hunk 1]
```diff
@@ -936,6 +936,10 @@ impl Document {
         point_in_node: Option<Point2D<f32>>,
         pressed_mouse_buttons: u16,
     ) {
+        if self.window.Closed() {
+            // If the BC is being discared, we don't want to run any JS.
+            return;
```
Comment (asajeffrey) [inline components/script/dom/document.rs]: If this is generally true, should we include it in codegen?

File: components/script/dom/htmliframeelement.rs [hunk 2]
```diff
@@ -668,15 +668,17 @@ impl VirtualMethods for HTMLIFrameElement {
...
                     exited_pipeline_id
                 );
-                exited_document
-                    .window()
-                    .window_proxy()
-                    .discard_browsing_context();
+                // We also need to start ignoring tasks now,
+                // rather than when the `PipelineExit` message arrives.
+                let exited_window = exited_document.window();
+                exited_window.ignore_all_tasks();
```
Comment (asajeffrey) [inline components/script/dom/htmliframeelement.rs]: Good  catch! Though possibly this should be part of `discard_browsing_context()`?

**Review by asajeffrey — COMMENTED** [review]
Yay, this looks like quite an improvement! A couple of minor questions, but otherwise this LGTM!

File: components/script/dom/document.rs (hunk 1 above)
Comment (gterzian) [reply components/script/dom/document.rs]: I'm not familiar with codegen, how would that work?

**Comment by gterzian** [comment]
@asajeffrey Ok I've tried bundling the `ignore_all_tasks` and `window_proxy.discard_browsing_context` into a single method on the window, let's see if that changes anything in the test results...

@bors-servo try=wpt

File: components/script/dom/htmliframeelement.rs (hunk 2 above)
Comment (gterzian) [reply components/script/dom/htmliframeelement.rs]: I'll try..

File: components/script/dom/document.rs (hunk 1 above)
Comment (gterzian) [reply components/script/dom/document.rs]: Note to self: fix the typo in "discarded"

**Comment by gterzian** [comment]
```
Layout thread disconnected.: "SendError(..)" (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(1) }, at src/libcore/result.rs:1084)
  │ stack backtrace:
  │    0: servo::main::{{closure}}::hf8b3c67cd313fc40 (0x55c21e1b2f8d)
  │    1: std::panicking::rust_panic_with_hook::h2e01ca08bfa50f6c (0x55c221442acc)
  │              at src/libstd/panicking.rs:481
  │    2: std::panicking::continue_panic_fmt::h89ac412ecc105abb (0x55c221442582)
  │              at src/libstd/panicking.rs:384
  │    3: rust_begin_unwind (0x55c221442476)
  │              at src/libstd/panicking.rs:311
  │    4: core::panicking::panic_fmt::hc72bd3ff9f8b21ef (0x55c22146493a)
  │              at src/libcore/panicking.rs:85
  │    5: core::result::unwrap_failed::h8fe51b7d2d9e1a4f (0x55c221464a37)
  │              at src/libcore/result.rs:1084
  │    6: script::dom::window::Window::force_reflow::h3f181f3a5f8c18a1 (0x55c21eceadf3)
  │    7: script::dom::window::Window::reflow::h162386a5db026bbd (0x55c21eceb547)
  │    8: script::dom::document::Document::finish_load::hdd48bd5ecca40f48 (0x55c21ef2d0cc)
  │    9: script::dom::servoparser::ServoParser::do_parse_sync::h04cb1a8f7b496d58 (0x55c21edec4b0)
  │   10: profile_traits::time::profile::h61da9b70f9113a3b (0x55c21f038f17)
  │   11: _ZN6script3dom11servoparser11ServoParser10parse_sync17h8071ded8c9797924E.llvm.11915649629530729400 (0x55c21edebb29)
  │   12: <script::dom::servoparser::ParserContext as net_traits::FetchResponseListener>::process_response_eof::ha886ddfd4d9f6244 (0x55c21edf1d0a)
  │   13: script::script_thread::ScriptThread::handle_msg_from_constellation::h585dc72d50e2e1fe (0x55c21efa8ab6)
  │   14: _ZN6script13script_thread12ScriptThread11handle_msgs17h9e8097a430f569ebE.llvm.1933365936389429086 (0x55c21efa0a09)
  │   15: script::script_thread::ScriptThread::start::h4842e8a890a7b74a (0x55c21ef9afb8)
  │   16: profile_traits::mem::ProfilerChan::run_with_memory_reporting::h1de62d815a33886b (0x55c21f0fd5bd)
  │   17: std::sys_common::backtrace::__rust_begin_short_backtrace::hceebb20f329f8edf (0x55c21e7635fe)
  │   18: _ZN3std9panicking3try7do_call17hbd11ebeb10fbe5a7E.llvm.3875094020376364713 (0x55c21f4b0264)
  │   19: __rust_maybe_catch_panic (0x55c22144c72a)
  │              at src/libpanic_unwind/lib.rs:80
  │   20: core::ops::function::FnOnce::call_once{{vtable.shim}}::hf2a26bab9b9abb71 (0x55c21e93ffa3)
  │   21: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::hf92c7ee34ebffad1 (0x55c22143113f)
  │              at /rustc/521d78407471cb78e9bbf47160f6aa23047ac499/src/liballoc/boxed.rs:922
  │   22: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::ha94d3a20f0e8dd18 (0x55c22144ba70)
  │              at /rustc/521d78407471cb78e9bbf47160f6aa23047ac499/src/liballoc/boxed.rs:922
  │       std::sys_common::thread::start_thread::h4ab30187e03381f3
  │              at src/libstd/sys_common/thread.rs:13
  │       std::sys::unix::thread::Thread::new::thread_start::hb63d4d1c8078a565
  │              at src/libstd/sys/unix/thread.rs:79
  │   23: start_thread (0x7f2d76aac6ba)
  │   24: clone (0x7f2d7534741d)
  │   25: <unknown> (0x0)
  │ [2019-09-10T06:09:22Z ERROR servo] Layout thread disconnected.: "SendError(..)"
  └ Pipeline failed in hard-fail mode.  Crashing!

{
    "status": "CRASH", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/resource-timing/nested-context-navigations-iframe.html", 
    "line": 286985, 
    "action": "test_result", 
    "expected": "TIMEOUT"
}
```

This one might be intermittent, and it's one of those "interesting" ones...

@bors-servo try=wpt

**Comment by jdm** [comment]
@bors-servo try=wpt
It's `try` for a new commit, and `retry` for a commit that hasn't changed since the last run.

**Comment by gterzian** [comment]
```
{
    "status": "FAIL", 
    "group": "default", 
    "message": "/css/css-values/vh-support-transform-translate.html b66c48d9abd17b3672f590a293821101a1847cb9\n/css/css-values/reference/all-green.html ca9048d25067fe2666ee98ffdaee4d41e63aa283", 
    "stack": null, 
    "subtest": null, 
    "test": "/css/css-values/vh-support-transform-translate.html", 
    "line": 38036, 
    "action": "test_result", 
    "expected": "PASS"
}
```

I think this one might be intermittent

**Comment by gterzian** [comment]
@asajeffrey ready for another round, see the last commit, to be squashed.

**Comment by CYBAI** [comment]
> I think this one might be intermittent

Yes, it's #23290.

**Comment by gterzian** [comment]
@asajeffrey This one is ready for another review.

**Comment by asajeffrey** [comment]
Yay, this looks like it's quite the improvement! @bors-servo r+

**Pull request merged**